pennylane/beta/plugins/default_qubit_tf.py

```diff
@@ -0,0 +1,442 @@
+            the expectation values. Defaults to 1000 if not specified.
+            If ``analytic == True``, then the number of shots is ignored
+            in the calculation of expectation values and variances, and only controls the number
+            of samples returned by ``sample``.
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+    """
+
+    name = "Default qubit (TensorFlow) PennyLane plugin"
+    short_name = "default.qubit.tf"
+    pennylane_requires = "0.10"
+    version = "0.10.0"
+    author = "Xanadu Inc."
+    _capabilities = {
+        "model": "qubit",
+        "provides_jacobian": False,
+        "passthru_interface": "tf",
+    }
+
+    operations = {
+        "BasisState",
+        "QubitStateVector",
+        "QubitUnitary",
+        "DiagonalQubitUnitary",
+        "PauliX",
+        "PauliY",
+        "PauliZ",
+        "MultiRZ",
+        "Hadamard",
+        "S",
+        "T",
+        "CNOT",
+        "SWAP",
+        "CSWAP",
+        "Toffoli",
+        "CZ",
+        "PhaseShift",
+        "RX",
+        "RY",
+        "RZ",
+        "Rot",
+        "CRX",
+        "CRY",
+        "CRZ",
+        "CRot",
+    }
+
+    observables = {"PauliX", "PauliY", "PauliZ", "Hadamard", "Hermitian", "Identity"}
+
+    parametric_ops = {
+        "PhaseShift": tf_ops.PhaseShift,
+        "RX": tf_ops.RX,
+        "RY": tf_ops.RY,
+        "RZ": tf_ops.RZ,
+        "Rot": tf_ops.Rot,
+        "CRX": tf_ops.CRX,
+        "CRY": tf_ops.CRY,
+        "CRZ": tf_ops.CRZ,
+    }
+
+    _asarray = staticmethod(tf.convert_to_tensor)
+
+    def __init__(self, wires, *, shots=1000, analytic=True):
+        # create the initial state
+        state = np.zeros(2 ** wires, dtype=np.complex128)
+        state[0] = 1
+        state = tf.convert_to_tensor(state, dtype=C_DTYPE)
+
+        # Internally, we store the state as a tensor of dimension [2]*wires
+        self._state = tf.reshape(state, [2] * wires)
+        self._pre_rotated_state = self._state
+
+        # call QubitDevice init
+        super().__init__(wires, shots, analytic)
+
+    def apply(self, operations, rotations=None, **kwargs):
+        rotations = rotations or []
+
+        # apply the circuit operations
+        for i, operation in enumerate(operations):
+
+            if i > 0 and isinstance(operation, (QubitStateVector, BasisState)):
+                raise DeviceError(
+                    "Operation {} cannot be used after other Operations have already been applied "
+                    "on a {} device.".format(operation.name, self.short_name)
+                )
+
+            self._apply_operation(operation)
+
+        # store the pre-rotated state
+        self._pre_rotated_state = self._state
+
+        # apply the circuit rotations
+        for operation in rotations:
+            self._apply_operation(operation)

```

Comment:
In terms of not having `QubitDevice` specifically apply rotations: we made this decision deliberately, since not all devices will want to implement rotations (e.g., there are some frameworks that have built-in support for measurements outside of just PauliZ)

For the state prep validation: I agree. The initial QubitDevice ADR made the decision to move this into the CircuitGraph (e.g., so the circuit graph will raise an error if the QNode attempts to initialize a circuit with multiple state preps), but it just hasn't been implemented yet. Although... could there be devices that allow for multiple state preps (like SF?)

Reply:
:laughing: looks like that feature was taken out of SF in a recent PR